### PR TITLE
Feature/10 maintenance mode

### LINKED ISSUE #10: Enable maintenance mode for ecosytem

### PR DESCRIPTION
**Pull request by Subaro**
Resolves #10

### CODE REVIEW

File: controllers/ingressUpdater_test.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	// then
+	assert.NotNil(t, creator)
+}
+
+func Test_ingressUpdater_UpdateIngressOfService(t *testing.T) {
+	t.Parallel()
+	myNamespace := "my-test-namespace"
+	myIngressClass := "my-ingress-class"
+
+	t.Run("skipped as service has no ports", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/ingressUpdater_test.go:34]: There are no ingress objects created as they are skipped

File: controllers/ingressUpdater_test.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+	// when
+	creator := NewIngressUpdater(clientMock, "my-namespace", "my-ingress-class-name")
+
+	// then
+	assert.NotNil(t, creator)
+}
+
+func Test_ingressUpdater_UpdateIngressOfService(t *testing.T) {
+	t.Parallel()
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/ingressUpdater_test.go:30]: I am not sure

File: controllers/ingressClassCreator.go [hunk 3]
```diff
@@ -16,67 +16,67 @@ import (
...
+	return &ingressClassCreator{
+		client:    client,
+		className: className,
 	}
 }
 
 // CreateIngressClass check whether the ingress class for the generator exists. If not it will be created.
-func (icc IngressClassCreator) CreateIngressClass(logger logr.Logger) error {
-	logger.Info(fmt.Sprintf("checking for existing ingress class [%s]", icc.ClassName))
+func (icc ingressClassCreator) CreateIngressClass(logger logr.Logger) error {
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/ingressClassCreator.go:34]: Yes

File: controllers/ingressUpdater.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+type ingressUpdater struct {
+	// client used to communicate with k8s.
+	client client.Client
+	// Namespace defines the target namespace for the ingress objects.
+	namespace string
+	// IngressClassName defines the ingress class for the ces services.
+	ingressClassName string
+}
+
+func NewIngressUpdater(client client.Client, namespace string, ingressClassName string) *ingressUpdater {
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/ingressUpdater.go:43]: Doine

File: controllers/ingressUpdater_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/ingressUpdater_test.go:30]: Yes. Done

File: controllers/maintenanceModeUpdater.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	log.FromContext(ctx).Info("Starting maintenance mode watcher...")
+	return scu.startEtcdWatch(ctx, scu.registry.RootConfig())
+}
+
+func (scu *maintenanceModeUpdater) startEtcdWatch(ctx context.Context, reg registry.WatchConfigurationContext) error {
+	log.FromContext(ctx).Info("Start etcd watcher on maintenance key")
+
+	warpChannel := make(chan *coreosclient.Response)
+	go func() {
+		log.FromContext(ctx).Info("start etcd watcher for maintenance key")
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/maintenanceModeUpdater.go:60]: Done

File: controllers/maintenanceModeUpdater.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+		reg.Watch(ctx, maintenanceModeWatchKey, true, warpChannel)
+		log.FromContext(ctx).Info("stop etcd watcher for maintenance key")
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/maintenanceModeUpdater.go:62]: Done

File: controllers/maintenanceModeUpdater_test.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+		err := maintenanceUpdater.Start(ctx)
+		cancelFunc()
+
+		// then
+		require.NoError(t, err)
+		mock.AssertExpectationsForObjects(t, regMock, watchContextMock, globalConfigMock, regMock)
+	})
+}
+
+func Test_maintenanceModeUpdater_activateMaintenanceMode(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/maintenanceModeUpdater_test.go:127]: Yeah they are unnecessary. I removed them. Done

File: controllers/serviceController.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	service := &corev1.Service{}
+	err := r.client.Get(ctx, req.NamespacedName, service)
+	if err != nil {
+		return &corev1.Service{}, fmt.Errorf("failed to get service: %w", err)
+	}
+
+	return service, nil
+}
+
+func (r *serviceReconciler) isMaintenanceModeActive() (bool, error) {
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/serviceController.go:95]: Yes :eye: I will remove one of them. Done.

File: controllers/suite_test.go [hunk 9]
```diff
@@ -69,23 +70,48 @@ var _ = BeforeSuite(func() {
...
-		IngressGenerator: ingressCreator,
-	}
-	err = reconciler.SetupWithManager(k8sManager)
+	// create ssl updater class
+	sslUpdater, err := NewSslCertificateUpdater(k8sManager.GetClient(), myNamespace)
+	Expect(err).ToNot(HaveOccurred())
+	err = k8sManager.Add(sslUpdater)
+	Expect(err).ToNot(HaveOccurred())
+
+	//// create warp menu creator
```
[the remark this answers is not in the document]
Comment (Subaro) [reply controllers/suite_test.go:99]: It does completely not work.

**Comment by nhinze23** [comment]
lgtm

**Pull request merged**